### PR TITLE
Modernize CI and Python tooling

### PR DESCRIPTION
**Pull request by palewire**
## Summary

Modernize django-bakery's packaging, tests, CI, release automation, documentation, and local developer tooling using the current python-open-source-template conventions.

This replaces setup.py and Pipenv with pyproject.toml and uv; adopts Ruff, ty, pytest, pre-commit, package verification, and a worktree-aware devcontainer; tests Python 3.11-3.14 with Django 5.2, 6.0, and 6.1; and moves PyPI/docs deployments to protected GitHub environments with OIDC.

## Related Issue

N/A

## Change Type

- [x] Bug fix
- [ ] Feature
- [x] Documentation
- [x] Maintenance
- [x] Breaking change

## Acceptance Criteria

- [x] The behavior matches the issue or request.
- [x] Relevant edge cases are covered.

## Validation

- `make verify`
- `make hooks`
- 79 tests across all 10 supported Python/Django combinations
- 82.19% branch coverage
- Independent code review, including regression fixes for dynamic `get_absolute_url` and `Path` build directories
- IAM policy validation and least-privilege simulation for the docs deployment role

## Documentation and Release Notes

- [x] Documentation is updated, or no update is needed.
- [x] `CHANGELOG.md` is updated for a user-facing, compatibility, or security change.

## Deployment setup

- PyPI trusted publishing is configured for the `pypi` environment.
- AWS OIDC is configured for the `docs-production` environment and limited to `palewire-docs/django-bakery/`.
- Legacy token/key secrets remain until the first OIDC deployments succeed.

### CODE REVIEW

File: bakery/static_views.py [hunk 1]
```diff
@@ -39,45 +57,50 @@
...
             continue
-        newpath = os.path.join(newpath, part).replace('\\', '/')
+        newpath = str(Path(newpath) / part).replace("\\", "/")
     if newpath and path != newpath:
-        return HttpResponseRedirect(newpath)
-    fullpath = os.path.join(document_root, newpath)
-    if os.path.isdir(fullpath) and default:
-        defaultpath = os.path.join(fullpath, default)
-        if os.path.exists(defaultpath):
+        return HttpResponseRedirect(f"/{quote(newpath, safe='/')}")
```
[the remark this answers is not in the document]
Comment (palewire) [reply bakery/static_views.py:73]: Fixed in 508b74e. Canonical redirects now always use a single leading slash and percent-encode the normalized path, so user input cannot produce a scheme-relative or external URL. The replacement CodeQL run passes, and `test_static_view_redirect_stays_on_origin` covers the behavior.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: bakery/views/detail.py [hunk 2]
```diff
@@ -25,58 +32,72 @@ class BuildableDetailView(DetailView, BuildableMixin):
...
+    def build_object(self, obj: object) -> None:
+        logger.debug("Building %s", obj)
         self.request = self.create_request(self.get_url(obj))
         self.set_kwargs(obj)
         target_path = self.get_build_path(obj)
         self.build_file(target_path, self.get_content())
 
-    def build_queryset(self):
+    def build_queryset(self) -> None:
         [self.build_object(o) for o in self.get_queryset().all()]
```
[the remark this answers is not in the document]
Comment (palewire) [reply bakery/views/detail.py]: Fixed in 0a50a77. This now uses a direct `for` loop, avoiding the unused list allocation while preserving build order and behavior.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+requires-python = ">=3.11"
+license = "MIT"
+license-files = ["LICENSE"]
+authors = [{ name = "Ben Welsh", email = "b@palewi.re" }]
+dependencies = [
+    "Django>=5.2,<6.2",
+    "boto3>=1.4.4",
+    "fs>=2.0.17",
+    "setuptools<81",
+]
```
[the remark this answers is not in the document]
Comment (palewire) [reply pyproject.toml:19]: Fixed in 0a50a77. The build-system constraint is now `setuptools>=80,<81`, aligned with the runtime pin required by PyFilesystem's `pkg_resources` import. Package builds, isolated wheel installation/import, and the full verification suite pass.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: bakery/management/commands/__init__.py [hunk 4]
```diff
@@ -1,102 +1,134 @@
...
 
+    key_list = list(keys)
     key_chunks = []
-    for i in range(0, len(keys), chunk_size):
+    for i in range(0, len(key_list), chunk_size):
         chunk = []
-        for key in (list(keys)[i:i+100]):
-            chunk.append({'Key': key})
+        for key in key_list[i : i + 100]:
+            chunk.append({"Key": key})
```
[the remark this answers is not in the document]
Comment (palewire) [reply bakery/management/commands/__init__.py:128]: Fixed in 0a50a77. Slicing now honors `chunk_size`, with a regression test proving five keys and a chunk size of two produce 2/2/1 delete batches.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: bakery/management/commands/publish.py [hunk 5]
```diff
@@ -1,89 +1,103 @@
...
-            help="Force a republish of all items in the build directory"
+            help="Force a republish of all items in the build directory",
         )
         parser.add_argument(
             "--dry-run",
             action="store_true",
             dest="dry_run",
             default="",
-            help="Display the output of what would have been uploaded removed, but without actually publishing."
+            help="Display the output of what would have been uploaded removed, but without actually publishing.",
```
[the remark this answers is not in the document]
Comment (palewire) [reply bakery/management/commands/publish.py:79]: Fixed in 0a50a77. Both `--force` and `--dry-run` now use boolean `False` defaults, with parser regression coverage.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: bakery/management/commands/unpublish.py [hunk 6]
```diff
@@ -1,29 +1,39 @@
...
+        command_options = cast("Mapping[str, object]", options)
+        aws_bucket_name = cast("str", command_options.get("aws_bucket_name"))
+        if aws_bucket_name:
+            aws_bucket_name = aws_bucket_name
         else:
             # Otherwise try to find it the settings
-            if not hasattr(settings, 'AWS_BUCKET_NAME'):
+            if not hasattr(settings, "AWS_BUCKET_NAME"):
                 raise CommandError(self.bucket_unconfig_msg)
             aws_bucket_name = settings.AWS_BUCKET_NAME
```
[the remark this answers is not in the document]
Comment (palewire) [reply bakery/management/commands/unpublish.py:36]: Fixed in 0a50a77. The no-op branch is removed; settings fallback now runs only when the command-line bucket name is empty.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).

File: bakery/views/dates.py [hunk 7]
```diff
@@ -68,65 +80,71 @@ class BuildableYearArchiveView(YearArchiveView, BuildableMixin):
...
         self.build_file(target_path, self.get_content())
 
-    def build_dated_queryset(self):
+    def build_dated_queryset(self) -> None:
         """
         Build pages for all years in the queryset.
         """
         qs = self.get_dated_queryset()
         years = self.get_date_list(qs)
         [self.build_year(dt) for dt in years]
```
[the remark this answers is not in the document]
Comment (palewire) [reply bakery/views/dates.py]: Fixed in 0a50a77. The year, month, and day archive builders now use direct loops. I also cleaned up the same side-effect-only pattern in serial build and publish paths.

> [!NOTE]
> Auto-replied by the [GitHub Copilot app](https://github.com/github/app).
